File: 프로그래머스/고득점 kit/해시/베스트앨범.cpp
#include <string>
#include <vector>
#include <iostream>
#include <map>
#include <algorithm>

using namespace std;

bool cmp1(const pair<string, int>& a, const pair<string, int>& b) {
	return a.second > b.second;
}

bool cmp2(const pair<int, int>& a, const pair<int, int>& b) {
	if (a.first == b.first)
		return a.second < b.second;
	return a.first > b.first;
}


vector<int> solution(vector<string> genres, vector<int> plays) {
	vector<int> answer;

	map<string, int> m;

	for (int i = 0; i < genres.size(); i++) {
		m[genres[i]] += plays[i];
	}

	int GenreCount = m.size();

	vector<pair<string, int>> v1;
	for (auto iter = m.begin(); iter != m.end(); iter++) {
		v1.push_back({ iter->first, iter->second });
	}

	sort(v1.begin(), v1.end(), cmp1);

	for (int j = 0; j < v1.size(); j++) {
		m[v1[j].first] = j;
	}

	///////////

	vector<pair<int, int>> arr[100];
	for (int i = 0; i < genres.size(); i++) {
		int inx = m[genres[i]];
		arr[inx].push_back({ plays[i], i });
	}


	for (int i = 0; i < GenreCount; i++) {
		sort(arr[i].begin(), arr[i].end(), cmp2);
	}

	for (int i = 0; i < GenreCount; i++) {
		if (arr[i].size() < 2) {
			answer.push_back(arr[i][0].second);
		}
		else {
			answer.push_back(arr[i][0].second);
			answer.push_back(arr[i][1].second);
		}
	}

	return answer;
}

File: 코틀린ps/프로그래머스/고득점 kit/해시/베스트앨범.kt
class Solution {
    fun solution(genres: Array<String>, plays: IntArray): IntArray {

        var mapGenre = mutableMapOf<String, Int>()

        for (i in 0..genres.size - 1) { // map 최초로 들어오는 key null check
            if (mapGenre[genres[i]] == null) mapGenre[genres[i]] = 0
            mapGenre[genres[i]] = mapGenre[genres[i]]!! + plays[i]
        }

        val listGenre = mapGenre.toList().sortedWith(compareByDescending { it.second })
        // sortedWith = 새로운 리스트 반환

        for (i in listGenre.indices) mapGenre[listGenre[i].first] = i

        val arr2 = Array(mapGenre.size) { ArrayList<Pair<Int, Int>>() } // <inx, value>
        // c++ 과 다르게 kotlin은 변수로 배열 사이즈 초기화 가능

        for (i in 0..genres.size - 1) {
            val inx = mapGenre[genres[i]]
            arr2[inx!!].add(Pair(i, plays[i]))  // Pair 추가 시 'Pair(first, second)' 형태!
        }


        for (i in 0..arr2.size - 1) arr2[i].sortByDescending { it.second }
        // 각 행(장르) 별로 정렬 수행

        val tempArrayList = arrayListOf<Int>()
        for (i in 0..arr2.size - 1) {
            tempArrayList.add(arr2[i][0].first)
            if (arr2[i].size >= 2) tempArrayList.add(arr2[i][1].first)
        }

        var answer = tempArrayList.toIntArray()  // ArrayList -> IntArray

        return answer
    }
}